### PR TITLE
DA-877 Add Metrics EHR endpoint

### PR DESCRIPTION
**Pull request by rennat**
still missing:

- site activity over time
- current site totals

### CODE REVIEW

**Comment by rennat** [comment]
## Some example rendered queries

### test_interval_day (test.unit_test.api_test.metrics_ehr_api_test.MetricsEhrApiIntervalTest) ...
```
SELECT interval_query.start_date,
       interval_query.end_date,

  (SELECT count(*) AS count_1
   FROM participant
   JOIN participant_summary ON participant.participant_id = participant_summary.participant_id
   WHERE participant_summary.hpo_id != :hpo_id_1
     AND (participant_summary.email IS NOT NULL
          OR participant_summary.email NOT LIKE :email_1)
     AND participant_summary.withdrawal_status = :withdrawal_status_1
     AND participant.is_ghost_id IS NOT true
     AND participant_summary.consent_for_electronic_health_records_time <= interval_query.start_date) AS consented_count,

  (SELECT count(*) AS count_1
   FROM participant
   JOIN participant_summary ON participant.participant_id = participant_summary.participant_id
   WHERE participant_summary.hpo_id != :hpo_id_1
     AND (participant_summary.email IS NOT NULL
          OR participant_summary.email NOT LIKE :email_1)
     AND participant_summary.withdrawal_status = :withdrawal_status_1
     AND participant.is_ghost_id IS NOT true
     AND participant_summary.ehr_receipt_time IS NOT NULL
     AND participant_summary.ehr_receipt_time <= interval_query.start_date) AS received_count
FROM
  (SELECT start_date_query.start_date AS start_date,
          date_add(start_date_query.start_date, interval 1 day) AS end_date
   FROM
     (SELECT calendar.day AS start_date
      FROM calendar
      GROUP BY calendar.day) AS start_date_query) AS interval_query
WHERE interval_query.start_date >= :start_date_1
  AND interval_query.start_date <= :start_date_2
ORDER BY interval_query.start_date
```

### test_interval_month (test.unit_test.api_test.metrics_ehr_api_test.MetricsEhrApiIntervalTest) ... 
```
SELECT interval_query.start_date,
       interval_query.end_date,

  (SELECT count(*) AS count_1
   FROM participant
   JOIN participant_summary ON participant.participant_id = participant_summary.participant_id
   WHERE participant_summary.hpo_id != :hpo_id_1
     AND (participant_summary.email IS NOT NULL
          OR participant_summary.email NOT LIKE :email_1)
     AND participant_summary.withdrawal_status = :withdrawal_status_1
     AND participant.is_ghost_id IS NOT true
     AND participant_summary.consent_for_electronic_health_records_time <= interval_query.start_date) AS consented_count,

  (SELECT count(*) AS count_1
   FROM participant
   JOIN participant_summary ON participant.participant_id = participant_summary.participant_id
   WHERE participant_summary.hpo_id != :hpo_id_1
     AND (participant_summary.email IS NOT NULL
          OR participant_summary.email NOT LIKE :email_1)
     AND participant_summary.withdrawal_status = :withdrawal_status_1
     AND participant.is_ghost_id IS NOT true
     AND participant_summary.ehr_receipt_time IS NOT NULL
     AND participant_summary.ehr_receipt_time <= interval_query.start_date) AS received_count
FROM
  (SELECT start_date_query.start_date AS start_date,
          date_add(start_date_query.start_date, interval 1 month) AS end_date
   FROM
     (SELECT str_to_date(date_format(calendar.day, :date_format_1), :str_to_date_1) AS start_date
      FROM calendar
      GROUP BY str_to_date(date_format(calendar.day, :date_format_1), :str_to_date_1)) AS start_date_query) AS interval_query
WHERE interval_query.start_date >= :start_date_1
  AND interval_query.start_date <= :start_date_2
ORDER BY interval_query.start_date
```

### test_interval_quarter (test.unit_test.api_test.metrics_ehr_api_test.MetricsEhrApiIntervalTest) ... 
```
SELECT interval_query.start_date,
       interval_query.end_date,

  (SELECT count(*) AS count_1
   FROM participant
   JOIN participant_summary ON participant.participant_id = participant_summary.participant_id
   WHERE participant_summary.hpo_id != :hpo_id_1
     AND (participant_summary.email IS NOT NULL
          OR participant_summary.email NOT LIKE :email_1)
     AND participant_summary.withdrawal_status = :withdrawal_status_1
     AND participant.is_ghost_id IS NOT true
     AND participant_summary.consent_for_electronic_health_records_time <= interval_query.start_date) AS consented_count,

  (SELECT count(*) AS count_1
   FROM participant
   JOIN participant_summary ON participant.participant_id = participant_summary.participant_id
   WHERE participant_summary.hpo_id != :hpo_id_1
     AND (participant_summary.email IS NOT NULL
          OR participant_summary.email NOT LIKE :email_1)
     AND participant_summary.withdrawal_status = :withdrawal_status_1
     AND participant.is_ghost_id IS NOT true
     AND participant_summary.ehr_receipt_time IS NOT NULL
     AND participant_summary.ehr_receipt_time <= interval_query.start_date) AS received_count
FROM
  (SELECT start_date_query.start_date AS start_date,
          date_add(start_date_query.start_date, interval 1 quarter) AS end_date
   FROM
     (SELECT date(concat(year(calendar.day), :concat_1, lpad((quarter(calendar.day) - :quarter_1) * :param_1 + :param_2, :lpad_1, :lpad_2), :concat_2)) AS start_date
      FROM calendar
      GROUP BY date(concat(year(calendar.day), :concat_1, lpad((quarter(calendar.day) - :quarter_1) * :param_1 + :param_2, :lpad_1, :lpad_2), :concat_2))) AS start_date_query) AS interval_query
WHERE interval_query.start_date >= :start_date_1
  AND interval_query.start_date <= :start_date_2
ORDER BY interval_query.start_date
```

### test_interval_week (test.unit_test.api_test.metrics_ehr_api_test.MetricsEhrApiIntervalTest) ... 
```
SELECT interval_query.start_date,
       interval_query.end_date,

  (SELECT count(*) AS count_1
   FROM participant
   JOIN participant_summary ON participant.participant_id = participant_summary.participant_id
   WHERE participant_summary.hpo_id != :hpo_id_1
     AND (participant_summary.email IS NOT NULL
          OR participant_summary.email NOT LIKE :email_1)
     AND participant_summary.withdrawal_status = :withdrawal_status_1
     AND participant.is_ghost_id IS NOT true
     AND participant_summary.consent_for_electronic_health_records_time <= interval_query.start_date) AS consented_count,

  (SELECT count(*) AS count_1
   FROM participant
   JOIN participant_summary ON participant.participant_id = participant_summary.participant_id
   WHERE participant_summary.hpo_id != :hpo_id_1
     AND (participant_summary.email IS NOT NULL
          OR participant_summary.email NOT LIKE :email_1)
     AND participant_summary.withdrawal_status = :withdrawal_status_1
     AND participant.is_ghost_id IS NOT true
     AND participant_summary.ehr_receipt_time IS NOT NULL
     AND participant_summary.ehr_receipt_time <= interval_query.start_date) AS received_count
FROM
  (SELECT start_date_query.start_date AS start_date,
          date_add(start_date_query.start_date, interval 1 week) AS end_date
   FROM
     (SELECT str_to_date(concat(yearweek(calendar.day), :concat_1), :str_to_date_1) AS start_date
      FROM calendar
      GROUP BY str_to_date(concat(yearweek(calendar.day), :concat_1), :str_to_date_1)) AS start_date_query) AS interval_query
WHERE interval_query.start_date >= :start_date_1
  AND interval_query.start_date <= :start_date_2
ORDER BY interval_query.start_date
```

File: rest-api/dao/metrics_ehr_service.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    start_date,
+    end_date,
+    site_ids,
+    interval_query
+  ):
+    common_subquery_where_arg = (
+      (ParticipantSummary.hpoId != TEST_HPO_ID)
+      & (
+        ParticipantSummary.email.isnot(None)
+        | sqlalchemy.not_(ParticipantSummary.email.like(TEST_EMAIL_PATTERN))
```
Comment (rennat) [inline rest-api/dao/metrics_ehr_service.py]: @Seeker1911 I based this query off of an existing one and I've been told that these `email` and `test HPO id` checks may no longer be needed now that we have the `isGhostId` field. Can you confirm?

File: rest-api/alembic/versions/671f14f9e358_add_ehr_metric_fields.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    globals()["downgrade_%s" % engine_name]()
+
+
+
+def upgrade_rdr():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.add_column('participant_summary', sa.Column('ehr_receipt_time', model.utils.UTCDateTime(), nullable=True))
+    op.add_column('participant_summary', sa.Column('ehr_status', model.utils.Enum(EhrStatus), nullable=True))
+    op.add_column('participant_summary', sa.Column('ehr_update_time', model.utils.UTCDateTime(), nullable=True))
+    # ### end Alembic commands ###
```
Comment (rennat) [inline rest-api/alembic/versions/671f14f9e358_add_ehr_metric_fields.py]: Should this migration also replace the existing `participant_view` to expose these new fields?

**Comment by Seeker1911** [comment]
If you pull in latest it will fix the CI build

File: rest-api/dao/metrics_ehr_service.py (hunk 1 above)
Comment (Seeker1911) [reply rest-api/dao/metrics_ehr_service.py]: Yes, A check for `isGhostId` will be more accurate for counts.

File: rest-api/participant_enums.py [hunk 3]
```diff
@@ -68,6 +68,11 @@ class SampleStatus(messages.Enum):
   QNS_FOR_PROCESSING = 17
   QUALITY_ISSUE = 18
 
+class EhrStatus(messages.Enum):
+  """Status of EHRs"""
+  MISSING = 0
```
Comment (Seeker1911) [inline rest-api/participant_enums.py]: I think MISSING could be misinterpreted in this context. Perhaps `NOT_PRESENT` ?

**Pull request merged**